File: src/lib/token-optimization/cache-layer.ts
/**
 * CacheLayer — namespaced LRU cache with TTL.
 *
 * Namespaces:
 *   clean     — cleaned input text
 *   summary   — extractive summaries
 *   embedding — text embeddings
 *   response  — final model responses
 *
 * Uses a simple FNV-1a hash for cache keys (no Web Crypto / Node needed).
 * LRU eviction removes the oldest 25 % of entries when the store is full.
 */

interface CacheEntry<T = unknown> {
  value: T;
  ts: number; // Date.now() ms
}

/** FNV-1a 32-bit hash — fast, deterministic, no external deps. */
function fnv1a(str: string): string {
  let hash = 2166136261;
  for (let i = 0; i < str.length; i++) {
    // eslint-disable-next-line no-bitwise
    hash = Math.imul(hash ^ str.charCodeAt(i), 16777619) >>> 0;
  }
  return hash.toString(16).padStart(8, "0");
}

export class CacheLayer {
  private readonly store = new Map<string, CacheEntry>();
  private hitCount  = 0;
  private missCount = 0;

  constructor(
    private readonly maxSize  = 1_000,
    private readonly ttlMs    = 3_600_000, // 1 hour
  ) {}

  // ------------------------------------------------------------------
  // Typed accessors
  // ------------------------------------------------------------------

  getClean(rawText: string): string | undefined       { return this.get("clean", rawText) as string | undefined; }
  setClean(rawText: string, cleaned: string): void    { this.set("clean", rawText, cleaned); }

  getSummary(text: string): string | undefined        { return this.get("summary", text) as string | undefined; }
  setSummary(text: string, summary: string): void     { this.set("summary", text, summary); }

  getEmbedding(text: string): number[] | undefined    { return this.get("embedding", text) as number[] | undefined; }
  setEmbedding(text: string, vec: number[]): void     { this.set("embedding", text, vec); }

  getResponse(key: string): unknown                   { return this.get("response", key); }
  setResponse(key: string, response: unknown): void   { this.set("response", key, response); }

  // ------------------------------------------------------------------
  // Generic get / set
  // ------------------------------------------------------------------

  get(namespace: string, content: string): unknown {
    const key   = this.cacheKey(namespace, content);
    const entry = this.store.get(key);

    if (!entry) { this.missCount++; return undefined; }
    if (Date.now() - entry.ts > this.ttlMs) {
      this.store.delete(key);
      this.missCount++;
      return undefined;
    }

    entry.ts = Date.now(); // refresh for LRU
    this.hitCount++;
    return entry.value;
  }

  set(namespace: string, content: string, value: unknown): void {
    const key = this.cacheKey(namespace, content);
    if (this.store.size >= this.maxSize && !this.store.has(key)) this.evict();
    this.store.set(key, { value, ts: Date.now() });
  }

  // ------------------------------------------------------------------
  // Stats & maintenance
  // ------------------------------------------------------------------

  get hits():    number { return this.hitCount; }
  get misses():  number { return this.missCount; }
  get size():    number { return this.store.size; }
  get hitRate(): number {
    const total = this.hitCount + this.missCount;
    return total ? Math.round((this.hitCount / total) * 10_000) / 10_000 : 0;
  }

  stats(): Record<string, unknown> {
    return {
      hits:     this.hitCount,
      misses:   this.missCount,
      hitRate:  this.hitRate,
      size:     this.store.size,
      maxSize:  this.maxSize,
      ttlMs:    this.ttlMs,
    };
  }

  clear(): void {
    this.store.clear();
    this.hitCount  = 0;
    this.missCount = 0;
  }

  // ------------------------------------------------------------------

  private cacheKey(namespace: string, content: string): string {
    return `${namespace}:${fnv1a(content)}`;
  }

  private evict(): void {
    const sorted = [...this.store.entries()].sort((a, b) => a[1].ts - b[1].ts);
    const nEvict = Math.max(1, Math.floor(sorted.length / 4));
    for (const [key] of sorted.slice(0, nEvict)) this.store.delete(key);
  }
}

File: middleware/cache_layer.py
"""
In-process LRU cache with TTL for cleaned inputs, summaries, embeddings,
and final model responses.

Thread-safety note: uses a plain dict — suitable for single-threaded or
cooperative async code.  Add threading.Lock wrapping if needed in
multi-threaded environments.
"""
from __future__ import annotations

import hashlib
import logging
import time
from typing import Any, Optional

logger = logging.getLogger(__name__)


class CacheLayer:
    """
    Namespaced, TTL-aware LRU cache.

    Namespaces used by the middleware:
      - ``clean``     : cleaned input text
      - ``summary``   : extractive summaries of conversation turns
      - ``embedding`` : text embeddings (list[float])
      - ``response``  : final model responses

    LRU eviction removes the oldest 25 % of entries when the cache is full.
    """

    def __init__(
        self,
        max_size: int = 1_000,
        ttl_seconds: float = 3_600.0,
    ) -> None:
        self._max_size = max_size
        self._ttl      = ttl_seconds
        # key → {"val": Any, "ts": float}
        self._store: dict = {}
        self._hits   = 0
        self._misses = 0

    # ------------------------------------------------------------------
    # Typed accessors (kept thin to remain testable)
    # ------------------------------------------------------------------

    def get_cleaned(self, raw_text: str) -> Optional[str]:
        return self._get("clean", raw_text)

    def set_cleaned(self, raw_text: str, cleaned: str) -> None:
        self._set("clean", raw_text, cleaned)

    def get_summary(self, text: str) -> Optional[str]:
        return self._get("summary", text)

    def set_summary(self, text: str, summary: str) -> None:
        self._set("summary", text, summary)

    def get_embedding(self, text: str) -> Optional[list]:
        return self._get("embedding", text)

    def set_embedding(self, text: str, embedding: list) -> None:
        self._set("embedding", text, embedding)

    def get_response(self, request_key: str) -> Optional[Any]:
        return self._get("response", request_key)

    def set_response(self, request_key: str, response: Any) -> None:
        self._set("response", request_key, response)

    # ------------------------------------------------------------------
    # Generic get / set
    # ------------------------------------------------------------------

    def get(self, namespace: str, content: str) -> Optional[Any]:
        return self._get(namespace, content)

    def set(self, namespace: str, content: str, value: Any) -> None:
        self._set(namespace, content, value)

    # ------------------------------------------------------------------
    # Stats & maintenance
    # ------------------------------------------------------------------

    @property
    def hits(self) -> int:
        return self._hits

    @property
    def misses(self) -> int:
        return self._misses

    @property
    def hit_rate(self) -> float:
        total = self._hits + self._misses
        return round(self._hits / total, 4) if total else 0.0

    @property
    def size(self) -> int:
        return len(self._store)

    def stats(self) -> dict:
        return {
            "hits":      self._hits,
            "misses":    self._misses,
            "hit_rate":  self.hit_rate,
            "size":      len(self._store),
            "max_size":  self._max_size,
            "ttl":       self._ttl,
        }

    def clear(self) -> None:
        self._store.clear()
        self._hits   = 0
        self._misses = 0

    # ------------------------------------------------------------------
    # Private helpers
    # ------------------------------------------------------------------

    def _cache_key(self, namespace: str, content: str) -> str:
        digest = hashlib.sha256(
            content.encode("utf-8", errors="replace")
        ).hexdigest()
        return f"{namespace}:{digest}"

    def _get(self, namespace: str, content: str) -> Optional[Any]:
        key   = self._cache_key(namespace, content)
        entry = self._store.get(key)

        if entry is None:
            self._misses += 1
            return None

        if time.monotonic() - entry["ts"] > self._ttl:
            del self._store[key]
            self._misses += 1
            return None

        # Refresh access time (LRU)
        entry["ts"] = time.monotonic()
        self._hits += 1
        return entry["val"]

    def _set(self, namespace: str, content: str, value: Any) -> None:
        key = self._cache_key(namespace, content)
        if len(self._store) >= self._max_size and key not in self._store:
            self._evict()
        self._store[key] = {"val": value, "ts": time.monotonic()}

    def _evict(self) -> None:
        """Remove the oldest 25 % of entries."""
        sorted_keys = sorted(self._store, key=lambda k: self._store[k]["ts"])
        n_evict = max(1, len(sorted_keys) // 4)
        for k in sorted_keys[:n_evict]:
            del self._store[k]
        logger.debug("CacheLayer: evicted %d entries (size now %d)", n_evict, len(self._store))
